File: ScaleformUI_Csharp/PauseMenu/Items/SettingsItems/SettingsListItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ScaleformUI.PauseMenu
{
    public delegate void SettingsListItemChanged(SettingsListItem item, int value, string listItem);
    public delegate void SettingsListItemSelected(SettingsListItem item, int value, string listItem);
    public class SettingsListItem : SettingsItem
    {
        private int itemIndex;
        public List<dynamic> ListItems { get; set; }
        public event SettingsListItemChanged OnListItemChanged;
        public event SettingsListItemSelected OnListItemSelected;
        public int ItemIndex
        {
            get => itemIndex;
            set
            {
                itemIndex = value;
                if (Parent != null)
                {
                    int tab = Parent.Parent.Parent.Tabs.IndexOf(Parent.Parent);
                    int leftItem = Parent.Parent.LeftItemList.IndexOf(Parent);
                    int rightIndex = Parent.ItemList.IndexOf(this);
                    Parent.Parent.Parent._pause.SetRightSettingsItemIndex(tab, leftItem, rightIndex, itemIndex);
                }
                ListChanged();
            }
        }

        public SettingsListItem(string label, List<dynamic> items, int startIndex) : base(label, "")
        {
            ItemType = SettingsItemType.ListItem;
            Label = label;
            ListItems = items;
            ItemIndex = startIndex;
        }

        public string CurrentItem()
        {
            return ListItems[ItemIndex].ToString();
        }

        public void ListSelected()
        {
            OnListItemSelected?.Invoke(this, itemIndex, CurrentItem());
        }

        public void ListChanged()
        {
            OnListItemChanged?.Invoke(this, itemIndex, "" + CurrentItem());
        }
    }
}

File: ScaleformUI_Lua/src/PauseMenu/Items/SettingsItems/SettingsListItem.lua
SettingsListItem = setmetatable({}, SettingsListItem)
SettingsListItem.__index = SettingsListItem
SettingsListItem.__call = function()
    return "SettingsItem", "SettingsItem"
end

function SettingsListItem.New(label, items, index)
    data = {
        Base = SettingsItem.New(label, ""),
        ItemType = SettingsItemType.ListItem,
        Label = label or "",
        ListItems = items or {},
        _itemIndex = index or 0,
        _enabled = true,
        _hovered = false,
        _selected = false,
        Parent = nil,
        OnListChanged = function(item, value, listItem)
        end,
        OnListSelected = function(item, value, listItem)
        end
    }
    return setmetatable(data, SettingsListItem)
end

function SettingsListItem:Enabled(enabled)
    if enabled ~= nil then
        self._enabled = enabled
        if self.Parent ~= nil and self.Parent.Parent ~= nil and self.Parent.Parent.Base.Parent ~= nil and self.Parent.Parent.Base.Parent:Visible() then
            if self.Parent:Selected() then
                local tab = IndexOf(self.Parent.Parent.Base.Parent.Tabs, self.Parent.Parent) - 1
                local leftItem = IndexOf(self.Parent.Parent.LeftItemList, self.Parent) - 1
                local rightIndex = IndexOf(self.Parent.ItemList, self) - 1
                ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("ENABLE_RIGHT_ITEM", false, tab, leftItem,
                rightIndex, self._enabled)
            end
        end
    else
        return self._enabled
    end
end

function SettingsListItem:Hovered(hover)
    if hover ~= nil then
        self._hovered = hover
    else
        return self._hovered
    end
end

function SettingsListItem:Selected(selected)
    if selected ~= nil then
        self._selected = selected
    else
        return self._selected
    end
end

function SettingsListItem:ItemIndex(index)
    if index ~= nil then
        self._itemIndex = index
        local tab = IndexOf(self.Parent.Parent.Base.Parent.Tabs, self.Parent.Parent) - 1
        local leftItem = IndexOf(self.Parent.Parent.LeftItemList, self.Parent) - 1
        local rightIndex = IndexOf(self.Parent.ItemList, self) - 1
        ScaleformUI.Scaleforms._pauseMenu:SetRightSettingsItemIndex(tab, leftItem, rightIndex, index)
        self.OnListChanged(self, self._itemIndex, tostring(self.ListItems[index]))
    else
        return self._itemIndex
    end
end
